rockcraft/services/lifecycle.py

```diff
@@ -0,0 +1,98 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Rockcraft Lifecycle service."""
+
+from __future__ import annotations
+
+import contextlib
+from pathlib import Path
+from typing import cast
+
+from craft_application import LifecycleService
+from craft_archives import repo
+from craft_cli import emit
+from craft_parts import callbacks
+from craft_parts.errors import CallbackRegistrationError
+from overrides import override
+
+from rockcraft.models.project import Project
+
+
+class RockcraftLifecycleService(LifecycleService):
+    """Rockcraft-specific lifecycle service."""
+
+    @override
+    def setup(self) -> None:
+        """Configure extra args to the LifecycleManager."""
+        # pylint: disable=import-outside-toplevel
+        # This inner import is necessary to resolve a cyclic import
+        from rockcraft.services import RockcraftServiceFactory
+
+        project = cast(Project, self._project)
+        project_vars = {"version": project.version}
+
+        services = cast(RockcraftServiceFactory, self._services)
+        image_service = services.image
+        image_info = image_service.obtain_image()
+
+        self._manager_kwargs.update(
+            base_layer_dir=image_info.base_layer_dir,
+            base_layer_hash=image_info.base_digest,
+            base=project.base,
+            package_repositories=project.package_repositories or [],
+            project_name=project.name,
+            project_vars=project_vars,
+        )
+
+        super().setup()
+
+    @override
+    def run(self, step_name: str, part_names: list[str] | None = None) -> None:
+        """Overridden to configure package repositories."""
```

Comment:
this feels more like a code comment, the override decorator takes care of the docstring (one of its uses)

Reply:
> the override decorator takes care of the docstring (one of its uses)

I thought so too but pydocstyle is complaining that the method doesn't have a docstring. I ended up duplicating the docstring and adding this as a code comment

Reply:
there's a required rule in pydocstyle for this to work, in setup.cfg style:
```
[pydocstyle]
# D107 Missing docstring in __init__ (reason: documented in class docstring)
# D203 1 blank line required before class docstring (reason: pep257 default)
# D213 Multi-line docstring summary should start at the second line (reason: pep257 default)
ignore = D107, D203, D213
ignore_decorators = overrides
```